### ISSUE TITLE
Why is `distance_mahalanobis` in correlation and not `effectsize`?

### ISSUE THREAD

**Issue by mattansb** [easystats/correlation#262]
I was going to implement Mahalanobis D in effectsize (as a multivariate Cohen's d) when I saw that this is here.

Functionally, this seems like a duplicate of check_outliers from performance. Can we remove this from here?

**Comment by IndrajeetPatil** [comment]
We are using it here: https://github.com/easystats/correlation/blob/7ad7b0050cd8dfcd7f713aacb5e85c2357cf6335/R/cor_test_shepherd.R#L6

You can move it to `{effectsize}`, and then we can just import it from there. 

`{effectsize}` doesn't rely on anything from `{correlation}`, right? Just to avoid any other circular dependency.

**Comment by mattansb** [comment]
This is _almost_ identical to `performance::check_outliers(method = "mahalanobis", threshold = 6)` or even just  `stats::mahalanobis() < 6` (see below), with the only difference being the bootstrap ("robust").

Since this function is being used only in this one place, I suggest:

1. Add the robust version to `check_outliers()`
    - (Maybe move `check_outliers.data.frame()` to `{datawizard}`?)
3. Remove this function completely and replace its function with the standard `mahalanobis() < 6` in `cor_test_shepherd()`.

(Or use `check_outliers()`, but `{performance}` currently suggests `{correlation}` (for a single line of code...), so this might be a circular dep problem?)

@DominiqueMakowski @bwiernik (past and present maintainers) - any objections?

``` r
set.seed(4)
some_data <- iris[,1:4]

d <- correlation::distance_mahalanobis(some_data)
not_outliers1 <- d$Distance < 6

outliers2 <- performance::check_outliers(some_data, method = "mahalanobis", threshold = 6)
not_outliers2 <- !outliers2

not_outliers3 <- mahalanobis(some_data, colMeans(some_data), cov(some_data)) < 6

cbind(not_outliers1, not_outliers2, not_outliers3, 
      diff = not_outliers1 != not_outliers2 | not_outliers1 != not_outliers3) |> 
  as.data.frame() |> 
  dplyr::arrange(-diff)
#>     not_outliers1 not_outliers2 not_outliers3  diff
#> 1           FALSE          TRUE          TRUE  TRUE
#> 2           FALSE          TRUE          TRUE  TRUE
#> 3           FALSE          TRUE          TRUE  TRUE
#> 4            TRUE          TRUE          TRUE FALSE
#> 5            TRUE          TRUE          TRUE FALSE
#> 6            TRUE          TRUE          TRUE FALSE
#> 7            TRUE          TRUE          TRUE FALSE
#> 8            TRUE          TRUE          TRUE FALSE
#> 9            TRUE          TRUE          TRUE FALSE
#> 10           TRUE          TRUE          TRUE FALSE
#> 11           TRUE          TRUE          TRUE FALSE
#> 12           TRUE          TRUE          TRUE FALSE
#> 13           TRUE          TRUE          TRUE FALSE
#> 14           TRUE          TRUE          TRUE FALSE
#> 15           TRUE          TRUE          TRUE FALSE
#> 16           TRUE          TRUE          TRUE FALSE
#> 17           TRUE          TRUE          TRUE FALSE
#> 18          FALSE         FALSE         FALSE FALSE
#> 19          FALSE         FALSE         FALSE FALSE
#> 20           TRUE          TRUE          TRUE FALSE
#> 21           TRUE          TRUE          TRUE FALSE
#> 22           TRUE          TRUE          TRUE FALSE
#> 23           TRUE          TRUE          TRUE FALSE
#> 24           TRUE          TRUE          TRUE FALSE
#> 25           TRUE          TRUE          TRUE FALSE
#> 26           TRUE          TRUE          TRUE FALSE
#> 27           TRUE          TRUE          TRUE FALSE
#> 28           TRUE          TRUE          TRUE FALSE
#> 29           TRUE          TRUE          TRUE FALSE
#> 30           TRUE          TRUE          TRUE FALSE
#> 31           TRUE          TRUE          TRUE FALSE
#> 32           TRUE          TRUE          TRUE FALSE
#> 33           TRUE          TRUE          TRUE FALSE
#> 34           TRUE          TRUE          TRUE FALSE
#> 35           TRUE          TRUE          TRUE FALSE
#> 36          FALSE         FALSE         FALSE FALSE
#> 37          FALSE         FALSE         FALSE FALSE
#> 38           TRUE          TRUE          TRUE FALSE
#> 39           TRUE          TRUE          TRUE FALSE
#> 40           TRUE          TRUE          TRUE FALSE
#> 41           TRUE          TRUE          TRUE FALSE
#> 42           TRUE          TRUE          TRUE FALSE
#> 43           TRUE          TRUE          TRUE FALSE
#> 44           TRUE          TRUE          TRUE FALSE
#> 45          FALSE         FALSE         FALSE FALSE
#> 46           TRUE          TRUE          TRUE FALSE
#> 47           TRUE          TRUE          TRUE FALSE
#> 48           TRUE          TRUE          TRUE FALSE
#> 49           TRUE          TRUE          TRUE FALSE
#> 50           TRUE          TRUE          TRUE FALSE
#> 51           TRUE          TRUE          TRUE FALSE
#> 52           TRUE          TRUE          TRUE FALSE
#> 53           TRUE          TRUE          TRUE FALSE
#> 54           TRUE          TRUE          TRUE FALSE
#> 55           TRUE          TRUE          TRUE FALSE
#> 56           TRUE          TRUE          TRUE FALSE
#> 57           TRUE          TRUE          TRUE FALSE
#> 58           TRUE          TRUE          TRUE FALSE
#> 59           TRUE          TRUE          TRUE FALSE
#> 60           TRUE          TRUE          TRUE FALSE
#> 61           TRUE          TRUE          TRUE FALSE
#> 62           TRUE          TRUE          TRUE FALSE
#> 63           TRUE          TRUE          TRUE FALSE
#> 64          FALSE         FALSE         FALSE FALSE
#> 65           TRUE          TRUE          TRUE FALSE
#> 66          FALSE         FALSE         FALSE FALSE
#> 67           TRUE          TRUE          TRUE FALSE
#> 68           TRUE          TRUE          TRUE FALSE
#> 69           TRUE          TRUE          TRUE FALSE
#> 70           TRUE          TRUE          TRUE FALSE
#> 71           TRUE          TRUE          TRUE FALSE
#> 72          FALSE         FALSE         FALSE FALSE
#> 73           TRUE          TRUE          TRUE FALSE
#> 74           TRUE          TRUE          TRUE FALSE
#> 75           TRUE          TRUE          TRUE FALSE
#> 76           TRUE          TRUE          TRUE FALSE
#> 77           TRUE          TRUE          TRUE FALSE
#> 78           TRUE          TRUE          TRUE FALSE
#> 79           TRUE          TRUE          TRUE FALSE
#> 80           TRUE          TRUE          TRUE FALSE
#> 81           TRUE          TRUE          TRUE FALSE
#> 82           TRUE          TRUE          TRUE FALSE
#> 83           TRUE          TRUE          TRUE FALSE
#> 84           TRUE          TRUE          TRUE FALSE
#> 85           TRUE          TRUE          TRUE FALSE
#> 86           TRUE          TRUE          TRUE FALSE
#> 87           TRUE          TRUE          TRUE FALSE
#> 88          FALSE         FALSE         FALSE FALSE
#> 89           TRUE          TRUE          TRUE FALSE
#> 90           TRUE          TRUE          TRUE FALSE
#> 91          FALSE         FALSE         FALSE FALSE
#> 92           TRUE          TRUE          TRUE FALSE
#> 93           TRUE          TRUE          TRUE FALSE
#> 94           TRUE          TRUE          TRUE FALSE
#> 95           TRUE          TRUE          TRUE FALSE
#> 96           TRUE          TRUE          TRUE FALSE
#> 97           TRUE          TRUE          TRUE FALSE
#> 98           TRUE          TRUE          TRUE FALSE
#> 99           TRUE          TRUE          TRUE FALSE
#> 100          TRUE          TRUE          TRUE FALSE
#> 101          TRUE          TRUE          TRUE FALSE
#> 102          TRUE          TRUE          TRUE FALSE
#> 103          TRUE          TRUE          TRUE FALSE
#> 104         FALSE         FALSE         FALSE FALSE
#> 105          TRUE          TRUE          TRUE FALSE
#> 106          TRUE          TRUE          TRUE FALSE
#> 107          TRUE          TRUE          TRUE FALSE
#> 108          TRUE          TRUE          TRUE FALSE
#> 109         FALSE         FALSE         FALSE FALSE
#> 110         FALSE         FALSE         FALSE FALSE
#> 111         FALSE         FALSE         FALSE FALSE
#> 112          TRUE          TRUE          TRUE FALSE
#> 113         FALSE         FALSE         FALSE FALSE
#> 114          TRUE          TRUE          TRUE FALSE
#> 115          TRUE          TRUE          TRUE FALSE
#> 116          TRUE          TRUE          TRUE FALSE
#> 117          TRUE          TRUE          TRUE FALSE
#> 118         FALSE         FALSE         FALSE FALSE
#> 119          TRUE          TRUE          TRUE FALSE
#> 120         FALSE         FALSE         FALSE FALSE
#> 121         FALSE         FALSE         FALSE FALSE
#> 122          TRUE          TRUE          TRUE FALSE
#> 123          TRUE          TRUE          TRUE FALSE
#> 124          TRUE          TRUE          TRUE FALSE
#> 125         FALSE         FALSE         FALSE FALSE
#> 126          TRUE          TRUE          TRUE FALSE
#> 127          TRUE          TRUE          TRUE FALSE
#> 128          TRUE          TRUE          TRUE FALSE
#> 129          TRUE          TRUE          TRUE FALSE
#> 130          TRUE          TRUE          TRUE FALSE
#> 131         FALSE         FALSE         FALSE FALSE
#> 132          TRUE          TRUE          TRUE FALSE
#> 133         FALSE         FALSE         FALSE FALSE
#> 134          TRUE          TRUE          TRUE FALSE
#> 135          TRUE          TRUE          TRUE FALSE
#> 136         FALSE         FALSE         FALSE FALSE
#> 137         FALSE         FALSE         FALSE FALSE
#> 138         FALSE         FALSE         FALSE FALSE
#> 139          TRUE          TRUE          TRUE FALSE
#> 140          TRUE          TRUE          TRUE FALSE
#> 141          TRUE          TRUE          TRUE FALSE
#> 142         FALSE         FALSE         FALSE FALSE
#> 143          TRUE          TRUE          TRUE FALSE
#> 144          TRUE          TRUE          TRUE FALSE
#> 145         FALSE         FALSE         FALSE FALSE
#> 146         FALSE         FALSE         FALSE FALSE
#> 147          TRUE          TRUE          TRUE FALSE
#> 148          TRUE          TRUE          TRUE FALSE
#> 149         FALSE         FALSE         FALSE FALSE
#> 150          TRUE          TRUE          TRUE FALSE
```

<sup>Created on 2022-09-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
Also @rempsyc

**Comment by rempsyc** [comment]
> Add the robust version to check_outliers()

There are already two robust versions of Mahalanobis in `performance::check_outliers()`, `method = mahalanobis_robust` (relies on [`bigutilsr::dist_ogk`](https://rdrr.io/cran/bigutilsr/man/covrob_ogk.html)), and `method = mcd` (relies on [`MASS::cov.mcd`](https://www.rdocumentation.org/packages/MASS/versions/7.3-58.1/topics/cov.rob)), but I'm pretty sure they don't use bootstrapping.

Would you like me to add the bootstrapping bit to the base `method = mahalanobis` in `check_outliers()` @mattansb? If so I'd like to ping you on #467 so we can merge that first.

> Maybe move check_outliers.data.frame() to {datawizard}?

Not sure about this. What would be the argument for doing that?

> (Or use check_outliers(), but {performance} currently suggests {correlation} (for a single line of code...), so this might be a circular dep problem?)

Hum, I see the problem. I think ideally we would remove the suggested `correlation` dependency in `performance`.

> Remove this function completely and replace its function with the standard mahalanobis() < 6 in cor_test_shepherd().

I have no objection to that, but what would be the pros/cons of doing this instead of using the current bootstrapped Mahalanobis? Breaking change? Regardless, that might be the easiest way to fix this if we don't need to change anything else given as you say the results are almost identical.

**Comment by mattansb** [comment]
My main issue is that this function is (1) not a correlation so shouldn't be here and (2) is used only once in the entirety of easystats codebase, so maybe just kill it? (and (3) I'm working on a [similarly named effectsize function](https://github.com/easystats/effectsize/pull/478) 🤠)

If we have a `performance::check_outliers(method = "mahalanobis_robust")` I don't think we need the bootstrapping.

> I have no objection to that, but what would be the pros/cons of doing this instead of using the current bootstrapped Mahalanobis? Breaking change? Regardless, that might be the easiest way to fix this if we don't need to change anything else given as you say the results are almost identical.

Pros: deals with points 1+2 above (and 3!), and makes the one use of the function more efficient if dropping the bootstrap.

Cons: Not sure. @DominiqueMakowski any thoughts? I think this was your code.

**Comment by mattansb** [comment]
Okay, I see what's going on here. This function is directly linked to Shepherd's pi (https://github.com/easystats/effectsize/pull/478), and Dom exported it.

So here is what I'm going to do - make it an internal function (unless @DominiqueMakowski objects).

**Comment by DominiqueMakowski** [comment]
https://github.com/easystats/correlation/blob/83d4203a134185ffbbea704018ff30f5f0690260/R/correlation.R#L78-L80

I don't have anything against moving that or making it an internal. The bootstrapped method was specifically described in the Shepherd's pi thing indeed. So if we are to re-implement Sheperd's pi with some other form of robust distance, it won't be the "original" one (it doesn't mean it would be worse though, bootstrapping is probably quite inefficient). 

In general, it would be sad to lose the bootstrap feature (even though it might not be optimal). Best would be to add it to check_outliers, and then to call it here, but we need to resolve the circular dep issue.

We could add a `get_distance()` function in datawizard | insight, though for now it seems like it would have little use to justify it (?)
